src/UCTSearch.cpp

```diff
@@ -200,14 +206,56 @@ SearchResult UCTSearch::play_simulation(GameState & currstate,
         if (currstate.get_passes() >= 2) {
             auto score = currstate.final_score();
             result = SearchResult::from_score(score);
+            // myprintf("double-pass %d\n", thread_num); ////
         } else if (m_nodes < MAX_TREE_SIZE) {
             float eval;
             const auto had_children = node->has_children();
-            const auto success =
-                node->create_children(m_network, m_nodes, currstate, eval,
-                                      get_min_psa_ratio());
+            
+            bool success;
+            if ((cfg_always_collect || collecting)) {
+                int rand_sym;
+                if (cfg_fixed_symmetry == -1) {
+                    rand_sym = Random::get_Rng().randfix<8>();
+                }
+                else {
+                    rand_sym = cfg_fixed_symmetry;
+                }
+                success = node->create_children(m_network, m_nodes, currstate, eval, get_min_psa_ratio(), rand_sym);
+                if (success && int( sym_states[color][thread_num].size() * cfg_num_threads ) < cfg_adj_positions) {
+                    auto sym_state = std::make_shared<Sym_State>();
+                    auto color = currstate.get_to_move();
+                    sym_state->symmetry = rand_sym;
+                    sym_state->state = currstate;
+                    sym_state->winrate = (color == FastBoard::BLACK ? eval : 1.0f - eval);
+                    //LOCK(get_mutex(), lock);
+                    sym_states[color][thread_num].push_back(sym_state);
+                    if ( int( (sym_states[color][thread_num].size() - 1) * cfg_num_threads ) >= cfg_adj_positions) {
+                        //sym_states[color][thread_num].pop_front();
+                    }
+                }
+            }
+            else {
+                success = node->create_children(m_network, m_nodes, currstate, eval, get_min_psa_ratio(), cfg_fixed_symmetry);
+            }
+
             if (!had_children && success) {
-                result = SearchResult::from_eval(eval);
+                if (currstate.eval_invalid() && !cfg_backup_fpu) {
+                    if (cfg_sure_backup) {
+                        auto state = currstate;
+                        do {
+                            currstate = state;
+                            result = play_simulation(currstate, node, thread_num);
+                        } while (!result.valid());
+                        result = SearchResult::from_eval(result.eval() + 2.0f);
```

Comment:
So what this is trying to do is to keep expanding until we have the right side to move?

Reply:
Yes. With the --pos or --neg option, the winrate (say) at black's turn may be adjusted but at white's turn may not be adjusted, so we can't use both winrates. With --pos or --neg it's also impossible to use net_eval for the fpu of a child, since the parent and the child have different sides to move. Hopefully after the network is properly trained, no --pos or --neg is necessary, and m_stm_komi, m_opp_komi (for NN inputs) and m_komi (for double-pass scoring) can always be set to be equal.